python/tank/platform/__init__.py

```diff
@@ -77,7 +77,42 @@ def _get_current_bundle():
 ################################################################################################
 # Public API methods
 
-def restart():
+def change_context(new_context):
+    """
+    Running change_context will attempt to change the context the engine and
+    its apps are running in on the fly. The current engine must accept the
+    context change, otherwise a full restart of the engine will be run instead.
+
+    Any apps that support context switching on the fly will do so. Any that do
+    not will be restarted.
+
+    The benefit of supporting context changes in engines and apps is speed. The
+    end result of this routine should be identical to that of a restart, but
+    will require less time to complete.
+
+    For more information on supporting context changing, see the following:
+
+        tank.platform.engine.context_change_allowed
+        tank.platform.application.context_change_allowed
+        tank.platform.engine.change_context()
+        tank.platform.application.change_context()
+    """
+    engine = current_engine()
+
+    if engine is None:
+        raise TankError("No engine is currently running! Run start_engine instead.")
+
+    if not isinstance(new_context, engine.context.__class__):
+        raise TankError("Given new context must be of type tank.context.Context.")
+
+    try:
+        engine.change_context(new_context)
```

Comment:
for debugging purposes, i think it could be nice to have a log message right before the `change_context()`call. That makes it easy to see (say in a support conversation) exactly when the change started and ended.